### ISSUE TITLE
HiveSplitter segfault with IceAct OMKey(0, 1) (Trac #1732)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1724]
Hi everyone,

HiveSplitter segfaults in HiveSplitter/private/HiveSplitter/HiveSplitter.cxx:522
The reason behind it is HiveSplitter/public/HiveSplitter/OMKeyHash.h:60 where a negative value is assigned to an unsigned int leading to a very large positive int. In line 522 it is tried to access uninitialised memory.

Probably it is best to tell HiveSplitter in the BOOST_FOR_EACH loop to ignore special DOMs like IceAct and the Scintillators checking omkey->IsScintillator() and omkey->IsIceAct() (the IceAct one does not exist yet)

Without a fix the muon level3 cannot be processed.

Cheers,
Leif

Migrated from https://code.icecube.wisc.edu/ticket/1732
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:12:47", 
    "description": "Hi everyone,\n\nHiveSplitter segfaults in HiveSplitter/private/HiveSplitter/HiveSplitter.cxx:522\nThe reason behind it is HiveSplitter/public/HiveSplitter/OMKeyHash.h:60 where a negative value is assigned to an unsigned int leading to a very large positive int. In line 522 it is tried to access uninitialised memory.\n\nProbably it is best to tell HiveSplitter in the BOOST_FOR_EACH loop to ignore special DOMs like IceAct and the Scintillators checking omkey->IsScintillator() and omkey->IsIceAct() (the IceAct one does not exist yet)\n\nWithout a fix the muon level3 cannot be processed.\n\nCheers,\nLeif", 
    "reporter": "lraedel", 
    "cc": "joertlin", 
    "resolution": "fixed", 
    "_ts": "1550067167842669", 
    "component": "combo reconstruction", 
    "summary": "HiveSplitter segfault with IceAct OMKey(0, 1)", 
    "priority": "blocker", 
    "keywords": "", 
    "time": "2016-06-08T19:45:45", 
    "milestone": "", 
    "owner": "mzoll", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20160608T21:09:10: 
* mzoll commented:

> Well, I kind of always said that IceHive/HiveSplitter is hardcoded to the IC86 geometry, so 86 strings with 64 DOMs each at maximum.
> 
> Because the hashing function is assuming this to be true, the OMKey(0,1) is not covered in this range.
> Then again I do not understand why IceAct is even handled with an OMKey, as it is obviously not a DOM and not at all comparable to the signals of the other DOMs. As I can kind of understand that this might have been necessary to do in order to handle the IceAct readout in a similar way to the triggers of the inice/icetop component, I am not quite sure if you should once you have the data north, continue transporting the IceAct information together with the I3RecoPulseSeriesMaps of the inice component (InIceRawData). Why don't you just remove/mask out this information from the Pulse-Series and have information of the inice component only transported to HiveSplitter/IceHive?
> 
> This would solve your problem and would represent the data as it was intended to.
> 
> If it should really really be necessary, I can also add the proposed check to the module but executing this check on something which should, if designed correctly, never ,raise this condition is kind of wasteful.
> 


* mzoll changed _comment0 from:

> Well, I kind of always aid that IceHive/HiveSplitter is hardcoded to the IC86 geometry, so 86 strings with 64 DOMs each at maximum.
> 
> Because th hashing function is assuming this to be true, the OMKey(0,1) is not covered in this range.
> Then again I do not understand why IceAct is even handled with an OMKey, as it is obviously not a DOM and not at all comparable to the signals of the other DOMs. As I can kind of aunderstand that this might have been neccessary to do in order to handle the IceAct readout in a similar way to the triggers of the inice/icetop component, I am not quite sure if you should once you have the data north keep on transporting the IceAct information together with the I3RecoPulseSeriesMaps of the inice componentn (InIceRawData). Why don't you just remove/mask out this information from the Pulse-Series and have information of the inice component onl transported to HiveSplitter/IceHive?
> 
> This would solve your problem and make would represent the data as it was intended to.
> 
> If it should really really be necessary, I can also add the proposed check to the module but executing this check on something which should if designed correctly never should raise this condition is kind of wasteful.
> 

to:

> 1465420286294227


* mzoll changed resolution from "" to "invalid"
* mzoll changed status from "new" to "closed"

**Comment by icecube-trac** [comment]
Trac update at 20160608T21:38:13: joertlin commented:

> I think we don't need to discuss if it is a good way to handle IceAct with OMKeys or not.
> 
> In my opinion a seg fault is the worst error you can get and it wastes time to debug it especially since we know what the cause of this error is. Therefore, I think the HiveSplitter should at least throw an error if a non in ice DOM appears. However, then we could also just skip it and print a warning.
> 
> Since it handles only in ice DOMs, you can just use `OMKey::IsInIce()` and you don't even need to implement the `IsIceAct()` method.

**Comment by icecube-trac** [comment]
Trac update at 20160608T22:17:57: mzoll commented:

> I agree that segfaults are a bad bad thing; however, if a software module is parsed something to that it is not designed to handle in any circumstances, and it tells you so in documentations etc. a segfault might be just be the result of that.
> 
> The thing is that HiveSplitter - as most of the icerec components known to me - is designed to act on the inice component OMKeys/DOMs only, and so does not explicit check if OMKeys passed to it are actually fulfilling this requirement.
> 
> That is exactly why the Pulseseries that such modules should operate on are derived from a key with the very descriptive name 'InIcePulses' established in BaseProcessing.
> 
> So again, I am am not questioning the choice of handling IceAct or any other special things by OMKeys, but rather the script and processing design to transport this information together with the regular inice component to modules where this information is not foreseen, and might it only be out of historical reasons.
> 
> If you really insist on this being the new way and this is sanctioned, you'd better check that every single module that you have down the line in the processing is doing the necessary checks and selections on the OMKeys too - something I honestly think they don't do for most cases!
> If this is then still needed I will implement that check.
> 
> But also understand this, even if the check of every OMKey seems a trivial thing to implement, this needs to be executed on every pulse that this module sees, and for IceHive running on uncleaned hits, that is a lot. And while a single call to this check might only take a few nanoseconds, this multiplied with a freakish huge number of pulses this adds up to quite a considerable number. Processing resources which can be better spend elsewhere, if the user would just have used the module as intended and specified to him...

**Comment by icecube-trac** [comment]
Trac update at 20160608T23:09:27: joertlin commented:

> If a module is designed just to handle InIce Keys, then it should take care of it that any other input throws at least reasonable errors, warnings or what ever but never never seg faults. I cannot understand why we should keep a seg fault especially since we can resolve it very easily.
> 
> I don't think that the additional need of computing power has any significant impact on the overall computing time even though `IsInIce()` needs to be called very often. I don't think we should sacrifice well error handling for micro-optimization.
> 
> Wasting time of people to debug seg faults doesn't make sense to me.

**Comment by icecube-trac** [comment]
Trac update at 20160609T05:25:43: lraedel commented:

> The segfault has absolutely nothing to do with the pulse map that is used. I don't think that IceAct is included in any pulse map since it actually doesn't really have a well defined waveform. In case it is that should not be the case. It should especially not be in the InIcePulses map.
> 
> The IceAct OMKey(0, 1) is in the geometry and therefore it is called in the mentioned for-loop which is called once for the configuration of HiveSplitter if I haven't missed something. Therefore, you need to check for the InIceDOMs when building the distance map. Since this is only called once there is also no real loss performance. 
> 
> Not sure why the scintillator OMs do not give problems, but this is something that would need to be checked since they also must not be included in the distance map.

**Comment by icecube-trac** [comment]
Trac update at 20160609T07:30:09: 
* lraedel changed resolution from "invalid" to ""
* lraedel changed status from "closed" to "reopened"

**Comment by icecube-trac** [comment]
Trac update at 20160609T14:45:48: olivas changed milestone from "" to "June 2016"

**Comment by icecube-trac** [comment]
Trac update at 20160609T15:57:28: 
* mzoll changed resolution from "" to "fixed"
* mzoll changed _comment0 from:

> In [changeset:"147172/IceCube"]:
> {{{
> #!CommitTicketReference repository="IceCube" revision="147172"
> fixes #1732 for good
> }}}

to:

> 1465487930869142


* mzoll commented:

> In changeset 147172:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="147172"
> fixes #1724 for good
> ```
> 
> HiveSplitter was alwas only designed for vanilla IC86. Please consider changing up to IceHive(Z)


* mzoll changed status from "reopened" to "closed"

**Comment by icecube-trac** [comment]
Trac update at 20160609T18:47:13: 
* olivas changed resolution from "fixed" to ""
* olivas commented:

> Looks like the fix http://code.icecube.wisc.edu/projects/icecube/changeset/147172/IceCube makes all the bots fail.


* olivas changed status from "closed" to "reopened"

**Comment by icecube-trac** [comment]
Trac update at 20160609T18:58:32: olivas commented:

> r147172 was backed out for the time being.

**Comment by icecube-trac** [comment]
Trac update at 20160610T07:20:35: 
* lraedel changed status from "reopened" to "closed"
* lraedel changed resolution from "" to "fixed"
* lraedel commented:

> In changeset 147228:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="147228"
> fixes #1724; revert changes by mzoll to fix this
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:12:47: 
* olivas commented:

> Milestone June 2016 deleted


* olivas changed milestone from "June 2016" to ""
